src/inference/graph_inference.cpp

```diff
@@ -883,35 +900,40 @@ class GraphNodeAssignment : public Nice2Assignment {
   void LocalPerFactorOptimizationPass(const GraphInference& fweights, size_t beam_size) {
     std::vector<std::pair<double, Factor>> empty;
     for (const Factor& factor : query_->factors_) {
-      std::vector<int> inf_vars;
-      inf_vars.reserve(factor.size());
+      std::vector<int> inf_nodes;
+      inf_nodes.reserve(factor.size());
       Factor giv_labels;
+      // Separates between given and to be inferred.
       for (auto var = factor.begin(); var != factor.end(); ++var) {
         Assignment a = assignments_[(*var)];
         if (a.must_infer == true) {
-          inf_vars.push_back((*var));
+          inf_nodes.push_back((*var));
         } else {
           giv_labels.insert(a.label);
         }
       }
-      if (inf_vars.size() <= 1) {
+      if (inf_nodes.size() <= 1) {
         continue;
       }
 
       std::vector<Factor> factors;
       GetFactorCandidates(fweights, factor.size(), &factors, giv_labels, beam_size);
       double best_score = 0;
-      std::vector<int> best_assignments;
-      best_assignments.assign(inf_vars.size(), 0);
-      for (size_t j = 0; j < inf_vars.size(); ++j) {
-        best_score += GetNodeScore(fweights, inf_vars[j]);
-        best_assignments[j] = assignments_[inf_vars[j]].label;
+      std::vector<int> best_assignments = std::vector<int>(inf_nodes.size());
```

Comment:
This is how you call the constructor.

std::vector<int> best_assignments(inf_nodes.size());